File: agent/clapcheeks/auth/token_vault.py
"""AI-8766 — Python side of the platform-token vault.

Wire-format compatible with ``web/lib/crypto/token-vault.ts``::

    byte 0       : version (currently 1)
    bytes 1..12  : iv (12 random bytes)
    bytes 13..28 : GCM auth tag (16 bytes)
    bytes 29..   : ciphertext

Per-user key = scrypt(MASTER_KEY, salt=user_id, n=16384, r=8, p=1, length=32).

The Python ``cryptography`` ``AESGCM`` helper concatenates ciphertext+tag and
hides the auth tag from the caller, which makes wire-compat with Node's
``crypto`` (which exposes ``getAuthTag()`` separately) awkward. We use the
lower-level ``Cipher`` API so we can place the tag exactly where Node puts
it.

Master key MUST be set via the ``CLAPCHEEKS_TOKEN_MASTER_KEY`` env var as a
base64-encoded 32-byte value (``openssl rand -base64 32``).
"""

from __future__ import annotations

import base64
import os
from typing import Final

from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives.kdf.scrypt import Scrypt

VERSION: Final[int] = 1
IV_BYTES: Final[int] = 12
TAG_BYTES: Final[int] = 16
KEY_BYTES: Final[int] = 32
HEADER_BYTES: Final[int] = 1 + IV_BYTES + TAG_BYTES  # 29

SCRYPT_N: Final[int] = 16384
SCRYPT_R: Final[int] = 8
SCRYPT_P: Final[int] = 1


class TokenVaultError(RuntimeError):
    """Raised when encryption / decryption fails for a recoverable reason
    (bad version, malformed blob, missing master key)."""


def _master_key() -> bytes:
    raw = os.environ.get("CLAPCHEEKS_TOKEN_MASTER_KEY")
    if not raw:
        raise TokenVaultError(
            "CLAPCHEEKS_TOKEN_MASTER_KEY not set. "
            "Generate with: openssl rand -base64 32"
        )
    raw = raw.strip().replace("-", "+").replace("_", "/")
    # Pad if missing trailing '='. base64.b64decode is strict about length.
    pad = "=" * (-len(raw) % 4)
    try:
        key = base64.b64decode(raw + pad)
    except Exception as exc:  # noqa: BLE001
        raise TokenVaultError(f"CLAPCHEEKS_TOKEN_MASTER_KEY is not valid base64: {exc}") from exc
    if len(key) != KEY_BYTES:
        raise TokenVaultError(
            f"CLAPCHEEKS_TOKEN_MASTER_KEY must decode to {KEY_BYTES} bytes, got {len(key)}"
        )
    return key


def _derive_key(user_id: str) -> bytes:
    if not user_id:
        raise TokenVaultError("user_id required for key derivation")
    kdf = Scrypt(
        salt=user_id.encode("utf-8"),
        length=KEY_BYTES,
        n=SCRYPT_N,
        r=SCRYPT_R,
        p=SCRYPT_P,
    )
    return kdf.derive(_master_key())


def encrypt_token(plaintext: str, user_id: str) -> bytes:
    """Encrypt ``plaintext`` (UTF-8 string) for ``user_id``.

    Returns wire-format bytes suitable for writing directly into a Postgres
    ``bytea`` column.
    """
    if not isinstance(plaintext, str):
        raise TokenVaultError("plaintext must be str")
    key = _derive_key(user_id)
    iv = os.urandom(IV_BYTES)
    encryptor = Cipher(algorithms.AES(key), modes.GCM(iv)).encryptor()
    ct = encryptor.update(plaintext.encode("utf-8")) + encryptor.finalize()
    tag = encryptor.tag
    if len(tag) != TAG_BYTES:
        raise TokenVaultError(f"unexpected GCM tag length {len(tag)}")
    return bytes([VERSION]) + iv + tag + ct


def decrypt_token(blob: bytes | bytearray | memoryview, user_id: str) -> str:
    """Decrypt a wire-format blob for ``user_id``.

    Raises ``TokenVaultError`` for recoverable issues (bad version,
    truncated blob). Raises ``cryptography.exceptions.InvalidTag`` if the
    blob was tampered with or encrypted under a different key.
    """
    buf = bytes(blob)
    if len(buf) < HEADER_BYTES + 1:
        raise TokenVaultError(f"blob too short ({len(buf)} bytes)")
    version = buf[0]
    if version != VERSION:
        raise TokenVaultError(f"unsupported vault version {version}")
    iv = buf[1 : 1 + IV_BYTES]
    tag = buf[1 + IV_BYTES : HEADER_BYTES]
    ct = buf[HEADER_BYTES:]
    key = _derive_key(user_id)
    decryptor = Cipher(algorithms.AES(key), modes.GCM(iv, tag)).decryptor()
    pt = decryptor.update(ct) + decryptor.finalize()
    return pt.decode("utf-8")


def decrypt_token_supabase(value, user_id: str) -> str | None:
    """Decrypt a value as returned by supabase-py for a ``bytea`` column.

    supabase-py / PostgREST return ``bytea`` columns as a string in one of
    two forms depending on configuration:

    * ``"\\xDEADBEEF..."`` — Postgres hex format
    * base64 — when the request includes the binary content-type header

    This helper handles both, plus the case where the column is already
    ``bytes`` (e.g. when called from psycopg directly). Returns ``None`` if
    the value is empty / falsy.
    """
    if value in (None, "", b""):
        return None
    if isinstance(value, (bytes, bytearray, memoryview)):
        return decrypt_token(value, user_id)
    if isinstance(value, str):
        if value.startswith("\\x"):
            return decrypt_token(bytes.fromhex(value[2:]), user_id)
        # Try base64; fall back to hex without prefix.
        try:
            return decrypt_token(base64.b64decode(value), user_id)
        except Exception:
            try:
                return decrypt_token(bytes.fromhex(value), user_id)
            except Exception as exc:  # noqa: BLE001
                raise TokenVaultError(
                    f"unrecognised bytea encoding for token: {exc}"
                ) from exc
    raise TokenVaultError(f"unsupported token value type: {type(value).__name__}")


__all__ = [
    "encrypt_token",
    "decrypt_token",
    "decrypt_token_supabase",
    "TokenVaultError",
    "VERSION",
    "IV_BYTES",
    "TAG_BYTES",
    "KEY_BYTES",
    "HEADER_BYTES",
]

File: web/lib/crypto/token-vault.ts
/**
 * AI-8766 — Token vault for Clapcheeks platform tokens.
 *
 * Provides application-level AES-256-GCM encryption / decryption for the
 * platform auth tokens (Tinder, Hinge, Bumble, Instagram) stored in
 * `clapcheeks_user_settings`. The encryption key is derived per-user from a
 * shared master key + the user_id (used as the scrypt salt) so:
 *
 *   - A leaked single-row blob does not reveal the master key.
 *   - Each user's blobs are independently keyed; cross-user replay is
 *     impossible even with master-key knowledge.
 *   - Master-key rotation = re-derive + re-encrypt all rows. The
 *     `token_enc_version` column is bumped to mark migrated rows.
 *
 * Wire format of an encrypted blob (Buffer):
 *
 *     byte 0       : version (currently 1)
 *     bytes 1..12  : iv (12 random bytes — GCM standard)
 *     bytes 13..28 : GCM auth tag (16 bytes)
 *     bytes 29..   : ciphertext
 *
 * This MUST stay byte-for-byte compatible with
 * `agent/clapcheeks/auth/token_vault.py`.
 */

import { createCipheriv, createDecipheriv, randomBytes, scryptSync } from 'crypto'

const VERSION = 1
const IV_BYTES = 12
const TAG_BYTES = 16
const KEY_BYTES = 32

const HEADER_BYTES = 1 + IV_BYTES + TAG_BYTES // 29

// scrypt parameters — must match the Python side.
// N=16384 (2**14), r=8, p=1 is the default the cryptography Python lib uses
// and remains the recommended interactive baseline (RFC 7914 / OWASP 2024).
const SCRYPT_N = 16384
const SCRYPT_R = 8
const SCRYPT_P = 1

function masterKey(): Buffer {
  const raw = process.env.CLAPCHEEKS_TOKEN_MASTER_KEY
  if (!raw) {
    throw new Error(
      'CLAPCHEEKS_TOKEN_MASTER_KEY not set. ' +
        'Generate with: openssl rand -base64 32',
    )
  }
  // Allow base64 OR base64url; trim whitespace/newlines.
  const cleaned = raw.trim().replace(/-/g, '+').replace(/_/g, '/')
  const buf = Buffer.from(cleaned, 'base64')
  if (buf.length !== KEY_BYTES) {
    throw new Error(
      `CLAPCHEEKS_TOKEN_MASTER_KEY must decode to ${KEY_BYTES} bytes, ` +
        `got ${buf.length}. Generate with: openssl rand -base64 32`,
    )
  }
  return buf
}

function deriveKey(userId: string): Buffer {
  if (!userId) {
    throw new Error('deriveKey: userId required (used as scrypt salt)')
  }
  return scryptSync(masterKey(), Buffer.from(userId, 'utf8'), KEY_BYTES, {
    N: SCRYPT_N,
    r: SCRYPT_R,
    p: SCRYPT_P,
    // scrypt's default maxmem is too small for N=16384; raise it.
    maxmem: 64 * 1024 * 1024,
  })
}

/**
 * Encrypt a UTF-8 string for `userId`.
 * Returns the wire-format Buffer. Caller writes it directly to a `bytea`
 * Postgres column.
 */
export function encryptToken(plaintext: string, userId: string): Buffer {
  if (typeof plaintext !== 'string') {
    throw new Error('encryptToken: plaintext must be a string')
  }
  const key = deriveKey(userId)
  const iv = randomBytes(IV_BYTES)
  const cipher = createCipheriv('aes-256-gcm', key, iv)
  const ct = Buffer.concat([
    cipher.update(plaintext, 'utf8'),
    cipher.final(),
  ])
  const tag = cipher.getAuthTag()
  if (tag.length !== TAG_BYTES) {
    throw new Error(`unexpected GCM tag length ${tag.length}`)
  }
  return Buffer.concat([Buffer.from([VERSION]), iv, tag, ct])
}

/**
 * Decrypt a wire-format blob for `userId`. Throws if the blob was encrypted
 * for a different user, with a different master key, or if it was tampered
 * with (GCM auth tag will fail).
 */
export function decryptToken(blob: Buffer | Uint8Array, userId: string): string {
  const buf = Buffer.isBuffer(blob) ? blob : Buffer.from(blob)
  if (buf.length < HEADER_BYTES + 1) {
    throw new Error(`decryptToken: blob too short (${buf.length} bytes)`)
  }
  const version = buf[0]
  if (version !== VERSION) {
    throw new Error(`decryptToken: unsupported vault version ${version}`)
  }
  const iv = buf.subarray(1, 1 + IV_BYTES)
  const tag = buf.subarray(1 + IV_BYTES, HEADER_BYTES)
  const ct = buf.subarray(HEADER_BYTES)
  const key = deriveKey(userId)
  const decipher = createDecipheriv('aes-256-gcm', key, iv)
  decipher.setAuthTag(tag)
  const pt = Buffer.concat([decipher.update(ct), decipher.final()])
  return pt.toString('utf8')
}

/**
 * Convenience for callers that already have a hex-encoded blob (e.g. read
 * from Supabase as the default \x-prefixed bytea representation).
 */
export function decryptTokenHex(hex: string, userId: string): string {
  // Postgres bytea hex format is `\x...`; strip the prefix if present.
  const clean = hex.startsWith('\\x') ? hex.slice(2) : hex
  return decryptToken(Buffer.from(clean, 'hex'), userId)
}

export const _internals = {
  VERSION,
  HEADER_BYTES,
  IV_BYTES,
  TAG_BYTES,
  KEY_BYTES,
  SCRYPT_N,
  SCRYPT_R,
  SCRYPT_P,
  deriveKey,
}
